client/transaction-pool/graph/src/validated_pool.rs

```diff
@@ -501,14 +502,53 @@ impl<B: ChainApi> ValidatedPool<B> {
 	pub fn status(&self) -> PoolStatus {
 		self.pool.read().status()
 	}
+
+	/// Notify all watchers that transactions in the block with hash have been finalized
+	pub async fn finalized(&self, block_hash: BlockHash<B>) -> Result<(), B::Error> {
+		debug!(target: "txpool", "Attempting to notify watchers of finalization for {}", block_hash);
+		let last_finalized_hash = *self.last_finalized_hash.lock();
+
+		let mut header = self.api.block_header(BlockId::Hash(block_hash.clone()))?
+			.expect("finalized block should have block header; qed");
+
+		while header.hash() != last_finalized_hash {
+			let hash = header.hash();
+			// fetch all extrinsic hashes
+			let tx_hashes = self.api.block_body(&BlockId::Hash(hash.clone())).await?
+				.expect("fetched block header should have block body; qed")
+				.into_iter()
+				.map(|tx| self.api.hash_and_length(&tx).0)
+				.collect::<Vec<_>>();
+
+			// notify the watcher that these extrinsics have been finalized
+			self.listener.write().finalized(&hash, &tx_hashes);
+
+			if last_finalized_hash == Default::default() {
+				// self.last_finalized_hash is created with a default value
+				break
+			}
+
+			header = self.api.block_header(BlockId::Hash(header.parent_hash().clone()))?
+				.expect("parent block should have block header; qed");
+		}
+
+		*self.last_finalized_hash.lock() = block_hash;
```

Comment:
This creates a race condition. It would probably best to fire this kind of notifications from a single-threaded environment or at least we need to fix the logic here.